aips/aip-x.md

```diff
@@ -0,0 +1,247 @@
+
+Another note is that this procedure is an on-chain procedure. Ideally we should make this a signed message instead so that users don't need to publish this intent to everyone.
+
+### Consuming intents
+```
+struct TradeSession<Args> {
+    argument: Args,
+    witness_type: TypeInfo,
+}
+
+public fun get_argument<Args>(session: &TradeSession<Args>): &Args {
+    &session.argument
+}
+
+public fun start_intent_session<Source: store, Args: store + drop>(
+    intent: Object<TradeIntent<Source, Args>>,
+): (Source, TradeSession<Args>)
+
+public fun finish_intent_session<Witness: drop, Args: store + drop>(
+    session: TradeSession<Args>,
+    _witness: Witness,
+)
+```
+There are two apis associated with claiming/consuming an on-chain intent. The first API will allow anyone to consume an intent object. Consumption of the intent object can release the resource locked in the intent temporarily. There will also be a hot potato type returned called `TradeSession`. Since this is a hot potato type, the transaction sender will have no choice but invoke the second `finish_intent_session` api which will require a value of `Witness` type. 
+
+### Example: Defining an intent to receive a certain fungible asset.
+```
+module aptos_framework::fungible_asset_intent_hooks {
+    struct FungibleAssetLimitOrder has store, drop {
+        desired_metadata: Object<Metadata>,
+        desired_amount: u64,
+        issuer: address,
+    }
+
+    struct FungibleAssetRecipientWitness has drop {}
+
+    public fun finish_fa_receiving_session(
+        session: TradeSession<FungibleAssetLimitOrder>,
+        received_fa: FungibleAsset,
+    ) {
+        let argument = intent::get_argument(&session);
+        assert!(
+            fungible_asset::metadata_from_asset(&received_fa) == argument.desired_metadata,
+            error::invalid_argument(ENOT_DESIRED_TOKEN)
+        );
+        assert!(
+            fungible_asset::amount(&received_fa) >= argument.desired_amount,
+            error::invalid_argument(EAMOUNT_NOT_MEET),
+        );
+
+        primary_fungible_store::deposit(argument.issuer, received_fa);
+        intent::finish_intent_session(session, FungibleAssetRecipientWitness {})
+    }
+}
+```
+Here we are defining an example of intent where the unlock condition is receiving a certain amount of a given fungible asset type. `Argument` type is instantiated to `FungibleAssetLimitOrder` and `Witness` type is instantated to `FungibleAssetRecipientWitness` in this example. In the arugment, we specify the amount and the type of FA we would like to receive. The check will then be implemented by first checking if the passed in `FungibleAsset` meets the requirement set in the `FungibleAssetLimitOrder` and deposit the FA to the issuer's primary fungible storage if the conditions are met. Once the checks are completed, we will provide the witness `FungibleAssetRecipientWitness` to complete this intent session.
+
```

Comment:
this needs an example of snippet of code of what the issuer and settler would need to do , for the above fungible_asset_intent_hooks.